packages/core/rsc/stream-ui/stream-ui.tsx

```diff
@@ -272,27 +272,85 @@ export async function streamUI<
     try {
       // Consume the forked stream asynchonously.
 
+      let textNodeIndex = 0;
       let content = '';
-      let hasToolCall = false;
+      let wasPreviouslyText = false;
+      const uiStreams = {} as Record<
+        string,
+        ReturnType<typeof createStreamableUI>
+      >;
+      let replaceInitial = !!initial;
 
       const reader = forkedStream.getReader();
       while (true) {
         const { done, value } = await reader.read();
+
+        if (wasPreviouslyText && (done || value.type !== 'text-delta')) {
+          const textNodeId = `text-${textNodeIndex}`;
+          if (!uiStreams[textNodeId]) {
+            throw new Error('Expected text node to exist');
+          }
```

Comment:
Is it possible to have text intermixed with tool calling?
I was able to get text _before_ tool calling, but trying to prompt at least the OpenAI agent to require a text response _after_ tool calling did not seem to work.

If text nodes can only appear once and **before** tool calling, this can be simplified instead of trying to keep track of nodes being written.

Reply:
Additionally, I noticed that with tool calling can come with a delay before the first chunk comes in for `reader.read()`. This had the unexpected effect of leaving out a last word or words of a text body that comes before tool calling and only shows up once the tool calls start streaming in.